### PR TITLE
Selective Intern

### PR DESCRIPTION
**Pull request by PaliC**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69043

**Selective Intern**

TL;DR This PR creates a shim for certain externed modules as described here https://fb.quip.com/I9PeAGfkt9Wl

Traditionally we extern entire libraries such as torch. However, sometimes parts of these libraries such as torch.fb or torch.quanitization are unstable. Therefore, this PR allows you to selectively intern parts of these externed submodules in order to maintain an internal copy of the unstable bits of the externed module. The primary use case for this is actually for torch which is externed automatically, but has unstable libraries such as torch.fb and torch.quantization which can cause breaks later on if they are externed.

We define a **selective extern** module as a module which we extern but there are submodules in it which we want to intern (ie. torch). A **selective intern** module is one of those submodules we want to intern (ie. torch.fb). 

In order to achieve this we added a new function to the package_exporter api _selective_intern(str package_name, [str] interned_packages) which makes package_name a selective extern module and those defined by interned_packages which are selectively interned modules. Selective externed modules return a shim when import is called on them which either returns a selectively interned submodule or an externed submodule when requested. We also extern all submodules which were not selective intern modules.

For example if we packaged something with 
```
package_exporter._selective_intern("torch",["torch.fb"])
```
and then run

```
import torch 
use(torch.fb)
use(torch.nn)
```
torch.fb would be interned and torch.nn would be externed.

Lastly, we still maintain that external dependencies of selective intern modules are still externed. For example if torch has a dependency on foo which is dependent on torch.fb, then foo's copy of torch.fb would be externed which calling torch.fb would still give you the interned module.

### CODE REVIEW

File: torch/nn/functional.py [hunk 1]
```diff
@@ -23,10 +23,10 @@
 from . import grad  # noqa: F401
 from .modules import utils
 from .modules.utils import _single, _pair, _triple, _list_with_default
-
+# import pdb
```
Comment (jbschlosser) [inline torch/nn/functional.py:26]: nit: please make sure the commented lines aren't merged :)

File: torch/nn/functional.py [hunk 2]
```diff
@@ -23,10 +23,10 @@
...
 
 Tensor = torch.Tensor
-
+# pdb.set_trace()
```
Comment (jbschlosser) [inline torch/nn/functional.py:29]: nit: please make sure the commented lines aren't merged :)
